File: com.trollworks.gcs/src/com/trollworks/gcs/ui/border/TitledBorder.java
/*
 * Copyright ©1998-2021 by Richard A. Wilkes. All rights reserved.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, version 2.0. If a copy of the MPL was not distributed with
 * this file, You can obtain one at http://mozilla.org/MPL/2.0/.
 *
 * This Source Code Form is "Incompatible With Secondary Licenses", as
 * defined by the Mozilla Public License, version 2.0.
 */

package com.trollworks.gcs.ui.border;

import com.trollworks.gcs.ui.GraphicsUtilities;
import com.trollworks.gcs.ui.TextDrawing;
import com.trollworks.gcs.ui.ThemeColor;
import com.trollworks.gcs.ui.ThemeFont;
import com.trollworks.gcs.ui.scale.Scale;

import java.awt.Color;
import java.awt.Component;
import java.awt.Font;
import java.awt.Graphics;
import java.awt.Graphics2D;
import java.awt.Insets;
import java.awt.Rectangle;
import javax.swing.SwingConstants;

/** A border consisting of a frame and optional title. */
public class TitledBorder extends LineBorder {
    private String    mTitle;
    private ThemeFont mFont;
    private Color     mTitleColor;

    /** Creates a new border without a title. */
    public TitledBorder() {
        super(ThemeColor.HEADER);
        mTitleColor = ThemeColor.ON_HEADER;
    }

    /**
     * Creates a new border.
     *
     * @param font  The font to use.
     * @param title The title to use.
     */
    public TitledBorder(ThemeFont font, String title) {
        this();
        mFont = font;
        mTitle = title;
    }

    /**
     * Creates a new border.
     *
     * @param borderColor The border color.
     * @param titleColor  The title color.
     * @param font        The font to use.
     * @param title       The title to use.
     */
    public TitledBorder(Color borderColor, Color titleColor, ThemeFont font, String title) {
        super(borderColor);
        mFont = font;
        mTitle = title;
        mTitleColor = titleColor;
    }

    /** @return The font. */
    public ThemeFont getFont() {
        return mFont;
    }

    /** @param font The font to use. */
    public void setFont(ThemeFont font) {
        mFont = font;
    }

    /** @return The title. */
    public String getTitle() {
        return mTitle;
    }

    /** @param title The new title. */
    public void setTitle(String title) {
        mTitle = title;
    }

    @Override
    public Insets getBorderInsets(Component component) {
        setThickness(Edge.TOP, (mTitle != null && mFont != null) ? TextDrawing.getPreferredSize(mFont.getFont(), mTitle).height : Scale.get(component).scale(1));
        return super.getBorderInsets(component);
    }

    @Override
    public void paintBorder(Component component, Graphics g, int x, int y, int width, int height) {
        super.paintBorder(component, g, x, y, width, height);
        if (mTitle != null && mFont != null) {
            Graphics2D gc        = GraphicsUtilities.prepare(g);
            Scale      scale     = Scale.get(component);
            Font       savedFont = gc.getFont();
            Font       font      = scale.scale(mFont.getFont());
            int        one       = scale.scale(1);
            gc.setFont(font);
            gc.setColor(mTitleColor);
            TextDrawing.draw(gc, new Rectangle(x + one, y, width - (one + one), TextDrawing.getPreferredSize(font, mTitle).height), mTitle, SwingConstants.CENTER, SwingConstants.TOP);
            gc.setFont(savedFont);
        }
    }

    public int getMinimumWidth(Component component) {
        Scale scale = Scale.get(component);
        if (mTitle != null && mFont != null) {
            return TextDrawing.getPreferredSize(scale.scale(mFont.getFont()), mTitle).width + scale.scale(1) * 4;
        }
        return scale.scale(1) * 4;
    }
}

File: ux/titled_border.go
// Copyright (c) 1998-2025 by Richard A. Wilkes. All rights reserved.
//
// This Source Code Form is subject to the terms of the Mozilla Public
// License, version 2.0. If a copy of the MPL was not distributed with
// this file, You can obtain one at http://mozilla.org/MPL/2.0/.
//
// This Source Code Form is "Incompatible With Secondary Licenses", as
// defined by the Mozilla Public License, version 2.0.

package ux

import (
	"github.com/richardwilkes/gcs/v5/model/colors"
	"github.com/richardwilkes/gcs/v5/model/fonts"
	"github.com/richardwilkes/toolbox/xmath"
	"github.com/richardwilkes/unison"
	"github.com/richardwilkes/unison/enums/filltype"
	"github.com/richardwilkes/unison/enums/paintstyle"
)

var _ unison.Border = &TitledBorder{}

// TitledBorder provides a titled line border.
type TitledBorder struct {
	Title string
	Font  unison.Font
}

func (t *TitledBorder) font() unison.Font {
	if t.Font == nil {
		return fonts.PageLabelPrimary
	}
	return t.Font
}

// Insets implements unison.Border
func (t *TitledBorder) Insets() unison.Insets {
	return unison.Insets{
		Top:    xmath.Ceil(t.font().LineHeight()) + 2,
		Left:   1,
		Bottom: 1,
		Right:  1,
	}
}

// Draw implements unison.Border
func (t *TitledBorder) Draw(gc *unison.Canvas, rect unison.Rect) {
	clip := rect.Inset(t.Insets())
	clip.Y += 0.5
	clip.Height -= 0.5
	path := unison.NewPath()
	path.SetFillType(filltype.EvenOdd)
	path.Rect(rect)
	path.Rect(clip)
	gc.DrawPath(path, colors.Header.Paint(gc, rect, paintstyle.Fill))
	text := unison.NewSmallCapsText(t.Title, &unison.TextDecoration{
		Font:            t.font(),
		OnBackgroundInk: colors.OnHeader,
	})
	text.Draw(gc, rect.X+(rect.Width-text.Width())/2, rect.Y+1+text.Baseline())
}
